File: passgen-app/passgen-web/src/main/java/com/passgen/core/actors/MarshallerActor.java
package com.passgen.core.actors;

import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.passgen.core.model.GeneratePassword;

import static com.passgen.core.model.GeneratePassword.withLength;

public class MarshallerActor extends UntypedActor {
    private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);

    private final ActorRef next;
    private final Gson gson = new Gson();

    public MarshallerActor(ActorRef next) {
        this.next = next;
    }

    @Override
    public void onReceive(Object m) throws Exception {
        if (m instanceof String) {
            LOG.debug("Unmarshalling message: {}", m);
            try {
                next.tell(gson.fromJson((String) m, GeneratePassword.class), getSender());
            } catch (JsonSyntaxException e) {
                getSelf().tell(Integer.parseInt((String) m), getSender());
            }
        } else if (m instanceof Integer) {
            next.tell(withLength((Integer) m), getSender());
        } else {
            unhandled(m);
        }
    }
}

File: passgen-app/passgen-scala/src/main/scala/com/passgen/core/actors/MarshallerActor.scala
package com.passgen.core.actors

import akka.actor._
import akka.event.slf4j.SLF4JLogging
import com.google.gson.Gson
import com.passgen.core.model.GeneratePassword

class MarshallerActor(next: ActorRef) extends Actor with SLF4JLogging {
  val gson = new Gson()
  def receive = {
    case m: String =>
      log.debug("Unmarshalling message: {}", m)
      next.tell(gson.fromJson(m, classOf[GeneratePassword]), sender())
    case m: Number =>
      log.debug("Unmarshalling message: {}", m)
      next.tell(GeneratePassword(m.intValue()), sender())
  }
}